sound/soc/sof/ipc.c

```diff
@@ -287,19 +287,34 @@ static void ipc_tx_next_msg(struct work_struct *work)
 		container_of(work, struct snd_sof_ipc, tx_kwork);
 	struct snd_sof_dev *sdev = ipc->sdev;
 	struct snd_sof_ipc_msg *msg;
+	int ret;
 
 	spin_lock_irq(&sdev->ipc_lock);
 
 	/* send message if HW read and message in TX list */
-	if (list_empty(&ipc->tx_list) || !snd_sof_dsp_is_ready(sdev))
+	if (list_empty(&ipc->tx_list) || !snd_sof_dsp_is_ipc_ready(sdev))
 		goto out;
 
 	/* send first message in TX list */
 	msg = list_first_entry(&ipc->tx_list, struct snd_sof_ipc_msg, list);
-	list_move(&msg->list, &ipc->reply_list);
-	snd_sof_dsp_send_msg(sdev, msg);
+	ret = snd_sof_dsp_send_msg(sdev, msg);
+	if (ret < 0) {
+
+		/*
+		 * if ipc tx fails, the msg will be retained in the msg_list,
+		 * so that it can be re-tried until it succeeds or times-out.
+		 */
+		dev_err_ratelimited(sdev->dev,
+				    "error: ipc tx failed with error %d\n",
```

Comment:
this error msg won't happen frequently?  why need use _ratelimited() here?

Reply:
@keyonjie I guess the rationale here because if it keeps on failing repeatedly and the ipc;s are being scheduled at a high rate, you're going to see the same message over and over.